src/main/java/hudson/tasks/junit/pipeline/JUnitResultsStepExecution.java

```diff
@@ -50,8 +52,9 @@ protected TestResultSummary run() throws Exception {
             TestResultAction testResultAction = JUnitResultArchiver.parseAndAttach(step, pipelineTestDetails, run, workspace, launcher, listener);
 
             if (testResultAction != null) {
-                // TODO: Once JENKINS-43995 lands, update this to set the step status instead of the entire build.
-                if (testResultAction.getResult().getFailCount() > 0) {
+                int testFailures = testResultAction.getResult().getResultByNode(nodeId).getFailCount();
+                if (testFailures > 0) {
+                    node.addOrReplaceAction(new WarningAction(testFailures + " tests failed"));
                     run.setResult(Result.UNSTABLE);
```

Comment:
Another option would be for `WarningAction.onAttached` to call `Run.setResult` automatically. On the other hand I can imagine wanting this drastic effect to be explicit.

Reply:
I wasn't familiar with `onAttached` previously, but I think you are referring to [`RunAction2.onAttached`](https://github.com/jenkinsci/jenkins/blob/7936d873a1550b060acc25de0c910aac0fcbebf7/core/src/main/java/jenkins/model/RunAction2.java#L40). We certainly could make some changes in `FlowNode` so that `onAttached` and `onLoad` would be called for actions implementing `RunAction2`, but as you mentioned I think this particular effect is best left explicit.